### PR TITLE
Rework the use of ParameterList in Ocean & THCM.

### PR DESCRIPTION
**Pull request by merijn**
Adds getDefaultParameters and getDefaultInitParameters for OMUSE to be able to
query the parameters it can use. Also introduces validation of the parameters
passed in.

This time with a considerably cleaner and simpler diff! I'm still updating/adding the tests, I will push them when I'm done

### CODE REVIEW

File: src/ocean/THCM.C [hunk 1]
```diff
@@ -828,6 +830,7 @@ THCM::THCM(Teuchos::ParameterList& params, Teuchos::RCP<Epetra_Comm> comm) :
         col_scaling->PutScalar(1.0);
     }
 
+    ReadParameters(paramList.sublist("Starting Parameters"));
```
Comment (Sbte) [inline src/ocean/THCM.C]: Why is this suddenly needed?

**Comment by merijn** [comment]
Yeah, I know, I already noticed something was wrong with the squashed version, but I wanted to get the tests working first before bothering to change/fix anything, and comparing the ParameterLists properly is trickier then I was hoping since haveSameValuesSorted isn't in the Trilinos version used on Travis.

After the lunch meeting I should have finished tests and then I will fix this commit.

**Comment by erik808** [comment]
>   Salinity Forcing = 1
>   Salinity Forcing = 0.1   [unused]

I probably compared different tests in the same directory, not sure

File: src/ocean/Ocean.C [hunk 2]
```diff
@@ -137,8 +135,9 @@ Ocean::Ocean(RCP<Epetra_Comm> Comm, Teuchos::ParameterList& oceanParamList)
 
     // Read starting parameters from xml
     Teuchos::ParameterList& startList =
-        thcmList.sublist("Starting Parameters");
-    THCM::Instance().ReadParameters(startList);
+        params_.sublist("THCM").sublist("Starting Parameters");
+
+    thcm_->setParameters(startList);
```
Comment (Sbte) [inline src/ocean/Ocean.C]: Can you put this change in a separate commit?

File: src/tests/test_parameterlist_oceandefault.C [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+//------------------------------------------------------------------
+namespace // local unnamed namespace (similar to static in C)
+{
+    RCP<Ocean> ocean;
+    RCP<Epetra_Comm>  comm;
+}
+
+//------------------------------------------------------------------
+TEST(ParameterListOcean, DefaultInitialization)
```
Comment (Sbte) [inline src/tests/test_parameterlist_oceandefault.C]: Can this be moved to the other file? THCM may complain (being a singleton), but it should still work as long as you don't try to use the old instance, right? Same for the THCM test files.

File: src/ocean/THCM.C [hunk 4]
```diff
@@ -219,44 +221,44 @@ THCM::THCM(Teuchos::ParameterList& params, Teuchos::RCP<Epetra_Comm> comm) :
...
-    comp_sal_int = paramList.get("Compute salinity integral", true);
+    hdim         = paramList.get<double>("Depth hdim");
+    double qz    = paramList.get<double>("Grid Stretching qz");
+    int  itopo   = paramList.get<int>("Topography");
+    bool flat    = paramList.get<bool>("Flat Bottom");
+    comp_sal_int = paramList.get<bool>("Compute salinity integral");
 
-    bool rd_mask = paramList.get("Read Land Mask", false); //== false in experiment0
+    bool rd_mask = paramList.get<bool>("Read Land Mask"); //== false in experiment0
+    std::string mask_file = paramList.get<std::string>("Land Mask");
```
Comment (Sbte) [inline src/ocean/THCM.C]: This should be inside of the if statement. If `rd_mask` is false it should be unused.

File: src/ocean/THCM.C [hunk 5]
```diff
@@ -267,10 +269,10 @@ THCM::THCM(Teuchos::ParameterList& params, Teuchos::RCP<Epetra_Comm> comm) :
         sres = 0;
     }
 
-    bool rd_spertm = paramList.get("Read Salinity Perturbation Mask", false);
+    bool rd_spertm = paramList.get<bool>("Read Salinity Perturbation Mask");
+    std::string spertm_file = paramList.get<std::string>("Salinity Perturbation Mask");
```
Comment (Sbte) [inline src/ocean/THCM.C]: Same here

**Comment by Sbte** [comment]
Oh and it would also be nice to have the THCM changes and Ocean changes in separate commits to make it easier to read and easier to debug if something breaks.

File: src/tests/test_parameterlist_oceandefault.C (hunk 3 above)
Comment (merijn) [reply src/tests/test_parameterlist_oceandefault.C]: I would prefer that, but when I initially had them in the same file, the Fortran code segfaulted and prints an error about variables already being initialised. If you know how I can make it work, I'm all ears.

File: src/ocean/THCM.C (hunk 4 above)
Comment (merijn) [reply src/ocean/THCM.C]: I had that initially, but then the test that nothing is unused fails (which I had to check that you guys get the right result).

Although I just realised a simple lifehack, I suppose I can just change the test parameter list to set "Read Land Mask" to true, then these *would* be use in the test...

Comment (Sbte) [reply src/ocean/THCM.C]: Yes, and if you really want to make sure `[unused]` also works you could set it to `false` and check that it's unused (but don't do the other check).
